### PR TITLE
On musl targets assume certain symbols exist (like pipe2 and accept4).

### LINKED ISSUE #56675: Musl pipe2 missing
Sorry for the cryptic title, didn't knew exactly what to put there. On musl targets (well `x86_64-unknown-linux-musl` and `arm-unknown-linux-musleabi`(hf)) for some reason pipe2 is not used (even on kernels that support it). Pipe2 is important to protect against race condition when multiple threads run external commands.

The following is an example that due to this bug will deadlock on the musl targets listed above (and probably all linux musl ones).

https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=f004af72249cff12f048b4ba16fc4d6e

Running with strace proves that pipe2 is not used, and that instead the fallback is:

```
[pid 11648] pipe( <unfinished ...>                                                                  
[pid 11647] ioctl(9, FIOCLEX <unfinished ...>                                                                                                                                                                     
[pid 11648] <... pipe resumed> [11, 12]) = 0                                                                                                                                                                       
[pid 11647] <... ioctl resumed> )       = 0                                                                            
[pid 11648] ioctl(11, FIOCLEX <unfinished ...>                                                         
[pid 11647] ioctl(10, FIOCLEX <unfinished ...>   
```

vs the `x86_64-unknown-linux-gnu` target:

```
[pid  2128] pipe2( <unfinished ...>                                                                                                                                                                               
[pid  2127] pipe2( <unfinished ...>                                                                                                                                                                               
[pid  2128] <... pipe2 resumed> [6, 7], O_CLOEXEC) = 0                                                                                                                                                            
[pid  2127] <... pipe2 resumed> [8, 9], O_CLOEXEC) = 0                
```

Versions affected:
- 1.30.1
- 1.31.0 - Current stable
- rustc 1.32.0-nightly (4a45578bc 2018-12-07)

Later Edit:
Just to be clear, if I compile a C file with the same musl toolchain that rust uses pipe2 works, so it's not a bug in Musl (or doesn't look like one to me).

### PR DESCRIPTION
**Pull request by adrian-budau**
This fixes #56675. 

I don't know if this is the best solution, or if I should also add some tests so I'm waiting for some feedback.

Thanks!

### CODE REVIEW

**Comment by adrian-budau** [comment]
This is what I tried initially, but the only solution I found for that was to hardcode every possible call of weak. The reason for this is that some symbols (like __pthread_get_minstack) are gnu specific and not present on musl.

I don’t think there is a way if a symbol exists in scope in rust (with a macro or other mechanism) to automate this.

**Comment by alexcrichton** [comment]
Ah right good point, and bummer! I wonder though if we could perhaps use the `syscall` function to get around this? That way we'd use pipe2 on musl I think as it wouldn't require a `dlsym`?

**Comment by adrian-budau** [comment]
Are you refering to runtime detection using syscall? That still wouldn’t work for the symbol I listed because it’s not a syscall, it’s from gnu libc. 

Or maybe I misunderstood you.

**Comment by alexcrichton** [comment]
Oh yeah that's what I mean for pipe2/accept4, I misunderstood the pthread one (thought that was only on glibc). For that yeah I think we're forced to use `#[cfg]`

**Comment by adrian-budau** [comment]
Ok, so I’ll change the weak macro to assume the symbol exists in the libc crate (since that’s the case for accept4/pipe2) and let the pthread function be under #[cfg]. Sounds okay like this?

**Comment by alexcrichton** [comment]
Seems plausible to me!

**Comment by alexcrichton** [comment]
Oops sorry I didn't see that there were updates here. I've been thinking a bit more about this and I think it's actually probably best if we leave the `weak!` macro as-is and use the `syscall` function in `libc` itself. @adrian-budau would you be up for switching to that for pipe2/accept4? I don't mind sending a PR as well!

**Comment by alexcrichton** [comment]
Oh I'll assume that [this comment](https://github.com/rust-lang/rust/issues/56675#issuecomment-448244426) was intended for here, so gonna reply here!

I forgot about other non-Linux platforms, sorry about that. The duplication is definitely one I think we'll want to avoid here! One thing I think we could do is have a `syscall!` macro perhaps which expands to `weak!` on non-Linux platforms and on Linux it expands to returning a Rust-defined function from `get` which uses `libc::syscall` under the hood. 

I feel like it's still probably best to use `syscall` directly if we can to avoid dynamic lookups and changing the binary requirements for musl in case anyone's playing around with an older implementation. How's that sound though?

**Comment by adrian-budau** [comment]
Sorry about that commnet, wasn't paying attention where I was posting.

Sounds good so far, but I also have a small change to your proposal. We can just make this `syscall!` macro return `ENOSYS` if `weak!` returns `None` on non-Linux unixes. This way we should have the least amount of code duplicated.

Also because we are using syscalls directly it means we are effectively giving up the emulation of system calls from musl, which is fine I guess because this is already being done in rust code.

**Comment by alexcrichton** [comment]
Sounds reasonable to me!

Yeah it's a bummer that we don't leverage the musl code too much, but the reality is that we end up doing the same thing anyway for other platforms like glibc, so we end up needing to vendor the logic :(

File: src/libstd/sys/unix/weak.rs [hunk 1]
```diff
@@ -77,3 +77,55 @@ unsafe fn fetch(name: &str) -> usize {
...
+    (fn $name:ident($($arg_name:ident: $t:ty),*) -> $ret:ty) => (
+        unsafe fn $name($($arg_name: $t),*) -> $ret {
+            use libc;
+
+            weak! { fn $name($($t),*) -> $ret }
+
+            if let Some(fun) = $name.get() {
+                fun($($arg_name),*)
+            } else {
+                libc::ENOSYS
```
Comment (alexcrichton) [inline src/libstd/sys/unix/weak.rs]: I think this'll have to return -1 and store ENOSYS in errno?

File: src/libstd/sys/unix/weak.rs [hunk 2]
```diff
@@ -77,3 +77,55 @@ unsafe fn fetch(name: &str) -> usize {
...
+                concat_idents!(SYS_, $name),
+                $(::sys::weak::SyscallParam::to_param($arg_name)),*
+            ) as $ret
+        }
+    )
+}
+
+#[cfg(target_os = "linux")]
+pub trait SyscallParam {
+    fn to_param(self) -> libc::c_long;
```
Comment (alexcrichton) [inline src/libstd/sys/unix/weak.rs]: Could this trait perhaps be eschewed in favor of `as libc::c_long` for each argument?

File: src/libstd/sys/unix/weak.rs (hunk 1 above)
Comment (adrian-budau) [reply src/libstd/sys/unix/weak.rs]: Yep, my bad, fixed!

File: src/libstd/sys/unix/weak.rs (hunk 2 above)
Comment (adrian-budau) [reply src/libstd/sys/unix/weak.rs]: Sorry for this, for some reason I didn't believe *mut T can be casted to c_long. Fixed now!

**Comment by alexcrichton** [comment]
@bors: r+

Thanks!

**Pull request merged**